### DISCUSSION TITLE
[Question]: Very short responses when getting completions from llama-cpp-python

### DISCUSSION THREAD

**Discussion by origintopleft** [danny-avila/LibreChat#873] [Q&A] [answered]
### What is your question?

I'm using LibreChat with the OpenAI endpoint, but instead of actual OpenAI, `OPENAI_REVERSE_PROXY` is pointed to the local system on port 8000, where llama-cpp-python is serving Llama 2 70B through an OpenAI compatible API. Mostly, this works. The only problem is the responses we get are very short.

Our previous chat UI was able to display messages of various lengths. Is there a setting I could change in order to allow longer responses?

### More Details

Deployed using docker compose from the git repo.

Docker version 24.0.5, from Ubuntu repositories
llama-cpp-python from Aug 17

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment, marked as answer]
Interesting, maybe max_tokens needs to be sent to the request, looks like the default for that project is 16 (which is incredibly low)? 

https://github.com/abetlen/llama-cpp-python/issues/542

Add a line after api\app\clients\OpenAIClient.js on line 64
`this.modelOptions.max_tokens = 2000;`

```js
    if (!this.modelOptions) {
      this.modelOptions = {
        ...modelOptions,
        model: modelOptions.model || 'gpt-3.5-turbo',
        temperature:
          typeof modelOptions.temperature === 'undefined' ? 0.8 : modelOptions.temperature,
        top_p: typeof modelOptions.top_p === 'undefined' ? 1 : modelOptions.top_p,
        presence_penalty:
          typeof modelOptions.presence_penalty === 'undefined' ? 1 : modelOptions.presence_penalty,
        stop: modelOptions.stop,
      };
    }
this.modelOptions.max_tokens = 2000; // new line
```

**Reply by origintopleft** [reply]
![ID: LibreChat screenshot. The user asked the sample question "How do I make an HTTP request in JavaScript?" and the Llama 2 backend returned a full length response, with examples using XMLHttpRequest and fetch.](https://github.com/danny-avila/LibreChat/assets/4407155/8e3048a0-b0ed-41f5-aa8e-e2693b7b5370)
Perfection. Thank you!

**Reply by ghost** [reply]
Hello, I've had the same problem while using a reverse OAI proxy. Would it be possible to add it in the config, or better, in the preset? Since it also affects the max possible context, sometimes it's better to have it at different values.

**Reply by danny-avila** [reply]
@Yardanico 

This is tricky because the API will fail if the token count of the messages and the max_tokens value goes over the context window.

In other words, max_tokens needs to be a variable or left empty so it can default to the limit on OpenAI's end, not to mention that our calculations can be off since we stream responses and don't get the official count from OpenAI, and there is variability between models.

```bash
  status: 400,
  json: {
    error: {
      message: "This model's maximum context length is 8192 tokens. However, you requested 16014 tokens (14 in the messages, 16000 in the completion). Please reduce the length of the messages or completion.",
      type: 'invalid_request_error',
      param: 'messages',
      code: 'context_length_exceeded'
    }
  }
```
it doesn't make sense to add it to presets when it will need to be omitted or vary in most user cases.

What I can do as a workaround is add a new env variable that will hardcode the value which will be useful for reverse proxies which I assume are not as strict as OpenAI with the `max_tokens` field

**Reply by ghost** [reply]
@danny-avila wait, does LibreChat not do message context trimming? Usually you'd tokenize all messages, and remove old messages that don't fit in the total context size (which is model_context_size - max_tokens - 3), of course excluding the things that always need to be there (e.g. the system prompt).

3 is often omitted in most apps because it's hard to notice, but it is required: OpenAI needs 3 tokens to start every response (this also assumes that max_tokens is > 0)

**Reply by danny-avila** [reply]
@Yardanico yes LibreChat does prune messages based on context similar to what you said, adapted from openai-cookbook: https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
```js
  getTokenCountForMessage(message) {
    let tokensPerMessage;
    let nameAdjustment;
    if (this.modelOptions.model.startsWith('gpt-4')) {
      tokensPerMessage = 3;
      nameAdjustment = 1;
    } else {
      tokensPerMessage = 4;
      nameAdjustment = -1;
    }
// ...rest
```

The extra tokens are the start/end token priming as shown in the cookbook.

What I was saying before is that setting max_tokens in the payload can introduce more API errors as the value will add itself to the current tokens to the api on top of the messages token count.

If my messages are 14 tokens, after already pruning/formatting messages for the API, but then I hardcode a value (let's say X) to max_tokens in the payload, it becomes 14 + X. It's better to omit `max_tokens` so that OpenAI can fill that value in for us, and we would never run into a mismatch issue between what we count for max_tokens and what they count

**Reply by danny-avila** [reply]
basically, we know that our count is going to be off because we're not getting the official count from OpenAI due to streaming, even if we are only off by 1 as shown here: https://github.com/danny-avila/LibreChat/issues/934

if we add another variable that adds tokens to our payload, we are adding another degree for error, allowing us to potentially fail by 2 different methods of generating a bad request.

**Reply by ghost** [reply]
The thing is - you *can* count normal messages accurately up to the exact token match as the API. For function calling you can't - but I assume that's not a big deal. Having no max_tokens is IMO worse than counting tokens and properly pruning message context, even if it's more complex.

**Reply by danny-avila** [reply]
omitting max_tokens is only an issue in your case, where it affects a reverse proxy's handling of response tokens; otherwise, OpenAI fills this in for us, so I don't see how it's worse.

I do see that the example cookbook I referenced has been updated since I've last seen it, so I will need to update my handling, but mind you, I don't believe we can objectively say that we can 'count normal messages accurately up to the exact token match as the API' when the cookbook itself has these warnings:
>Below is an example function for counting tokens for messages passed to gpt-3.5-turbo or gpt-4.
> Note that the exact way that tokens are counted from messages may change from model to model. Consider the counts from the function below an estimate, not a timeless guarantee.

```python
    elif "gpt-3.5-turbo" in model:
        print("Warning: gpt-3.5-turbo may update over time. Returning num tokens assuming gpt-3.5-turbo-0613.")
# ...
    elif "gpt-4" in model:
        print("Warning: gpt-4 may update over time. Returning num tokens assuming gpt-4-0613.")
```

**Reply by danny-avila** [reply]
Note that I am doing a fair amount of token handling already, so I am taking into account the "remaining" tokens which is essentially what `max_tokens` represents and counting tokens as accurately as I know how (which I admitted needs an update). My qualm is with sending this "remaining tokens value" to the API, which is not necessary, unless you explicitly want shorter response than the remaining tokens or max_tokens possible. Most users want longer responses not shorter, and i hope to mediate the shorter response desire with 'stop' tokens in presets.

I concede the benefit in adding max_tokens as a 'preset' parameter when you intentionally want responses shorter than the current context window. technically, this is already supported in my token handling logic and would not be hard to add, and I would send it to the API in this case, but not by default.

**Reply by danny-avila** [reply]
@Yardanico thanks for this discussion earlier, helped me revisit what exactly I was doing wrong in the token count. Have added new tests and I'm matching the same count as updated cookbook example: https://github.com/danny-avila/LibreChat/pull/945

**Reply by danny-avila** [reply]
some good discussion on `max_tokens` here: https://github.com/vllm-project/vllm/pull/852

**Comment by danny-avila** [comment]
You can also try editing this line in llama-cpp-python 

https://github.com/abetlen/llama-cpp-python/blob/18337267c175f88c46fbad079a7c18da57e1b520/llama_cpp/llama.py#L863

**Reply by origintopleft** [reply]
Typically on the llama-cpp-python side, settings like that are exposed as environment variables. Or at least that's what I used the last couple of times I had to change a llama setting. If it's not already exposed there, I'll see about getting a pull request started on their end.
